### PR TITLE
[WIP] HEARTBEAT and commander failsafe handling cleanup

### PR DESCRIPTION
**Pull request by dagar**
Incremental piece of https://github.com/PX4/Firmware/pull/10933.

### Testing
 - [x] data link loss & regain (GCS HEARTBEAT) : done https://github.com/Auterion/Firmware/tree/pr-heartbeat_cleanup
 - [ ] onboard computer link loss & regain (HEARTBEAT)
 - [ ] iridium loss & regain (slightly more difficult)

### CODE REVIEW

**Comment by thomasgubler** [comment]
@cmic0  will test above points in sitl

**Comment by MaEtUgR** [comment]
As just discussed in a call:
This needs to be reviewed, tested and merged as a first step to enable a fusion of #11457 and #11456 to have failsafe features when loosing the external obstacle avoidance module connected via MAVLink.

File: msg/vehicle_status.msg [hunk 1]
```diff
@@ -71,10 +72,13 @@ bool rc_signal_lost				# true if RC reception lost
 uint8 rc_input_mode				# set to 1 to disable the RC input, 2 to enable manual control to RC in mapping.
 
 bool data_link_lost				# datalink to GCS lost
-bool high_latency_data_link_active 		# all low latency datalinks to GCS lost
 uint8 data_link_lost_counter			# counts unique data link lost events
+
+bool high_latency_data_link_lost 			# all low latency datalinks to GCS lost
```
Comment (dagar) [inline msg/vehicle_status.msg]: I think these make more sense merged in with vehicle_status_flags.

eg
 - condition_data_link_valid
 - condition_avoidance_system_valid
 - condition_manual_control_valid

**Comment by dagar** [comment]
Rebased on master.

File: src/modules/commander/Commander.cpp [hunk 2]
```diff
@@ -3872,203 +3813,117 @@ bool Commander::preflight_check(bool report)
...
-					_telemetry[i].last_heartbeat = hrt_absolute_time();
-				}
-			}
-
-			if (!status.data_link_lost) {
-				mavlink_log_critical(&mavlink_log_pub, "ALL LOW LAT LINKS LOST, ACTIVATE HIGH LAT LINK");
+	// ONBOARD CONTROLLER data link loss failsafe (hard coded 5 seconds)
+	//  only issue a periodic warning for now
+	if ((_datalink_last_heartbeat_onboard_controller > 0)
+	    && (hrt_elapsed_time(&_datalink_last_heartbeat_onboard_controller) > 5_s) &&
```
Comment (dagar) [inline src/modules/commander/Commander.cpp]: TODO: review hardcoded 5 seconds.

File: src/modules/commander/Commander.hpp [hunk 3]
```diff
@@ -155,27 +166,21 @@ class Commander : public ModuleBase<Commander>, public ModuleParams
...
-		uint64_t last_heartbeat = 0u;
-		uint64_t last_dl_loss = 0u;
-		bool preflight_checks_reported = false;
-		bool lost = true;
-		bool high_latency = false;
-	} _telemetry[ORB_MULTI_MAX_INSTANCES];
+	void		data_link_check(bool &status_changed);
+	int		_telemetry_status_sub[ORB_MULTI_MAX_INSTANCES] {};
+
+	uint64_t	_datalink_last_heartbeat_gcs{0};
```
Comment (dagar) [inline src/modules/commander/Commander.hpp]: TODO: simplify naming and make consistent, we might end up with a longer list of types to handle.

Comment (dagar) [reply src/modules/commander/Commander.hpp]: hrt_abstime _heartbeat_gcs_last

File: src/modules/commander/Commander.hpp [hunk 4]
```diff
@@ -155,27 +166,21 @@ class Commander : public ModuleBase<Commander>, public ModuleParams
...
+	uint64_t	_datalink_lost{0};
+
+	uint64_t	_datalink_last_heartbeat_onboard_controller{0};
```
Comment (dagar) [inline src/modules/commander/Commander.hpp]: hrt_abstime _heartbeat_onboard_controller_last

File: src/modules/commander/Commander.cpp [hunk 5]
```diff
@@ -3872,203 +3813,121 @@ bool Commander::preflight_check(bool report)
...
 
-				if (isbd_timestamp > _telemetry[i].last_heartbeat) {
-					_telemetry[i].last_heartbeat = isbd_timestamp;
-				}
-			}
-		}
-	}
-}
+					// lazily subscribe
+					if (orb_exists(ORB_ID(iridiumsbd_status), 0) == PX4_OK) {
```
Comment (cmic0) [inline src/modules/commander/Commander.cpp]: I think you should use `if (_iridiumsbd_status_sub == -1 && orb_exists(ORB_ID(iridiumsbd_status), 0) == PX4_OK)` to prevent subscribing multiple times

**Comment by cmic0** [comment]
Tested **data link loss & regain (GCS HEARTBEAT)**. Did some fixes suggestions here: https://github.com/Auterion/Firmware/tree/pr-heartbeat_cleanup

**Comment by cmic0** [comment]
In order to test **onboard computer link loss & regain (HEARTBEAT)** and then fuse  #11457 and #11456 @baumanta commits from [84c0863](https://github.com/PX4/Firmware/pull/10933/commits/84c0863976406f391da4f74dce8014be3c4cc817) to [99f8e38](https://github.com/PX4/Firmware/pull/10933/commits/99f8e38de02805723ce725c4299e35f5ee8cc4fa) need be brought over here.

FYI @MaEtUgR @dagar @mrivi

File: msg/vehicle_status.msg (hunk 1 above)
Comment (cmic0) [reply msg/vehicle_status.msg]: Looks good to me. Can we assume this will be the final naming convention so that [#11457](https://github.com/PX4/Firmware/pull/11457) and [#11456](https://github.com/PX4/Firmware/pull/11456) can be adapted?
FYI @mrivi @MaEtUgR

Comment (mrivi) [reply msg/vehicle_status.msg]: @cmic0 fine for me

**Comment by thomasgubler** [comment]
Discussed with @cmic0 offline:
He is currently working at this branch https://github.com/Auterion/Firmware/tree/pr-heartbeat_cleanup
On that branch "data link loss & regain" is fixed and tested
Next he will add @baumanta 's commits on top and fix if necessary to complete  "onboard computer link loss & regain (HEARTBEAT)"
then he will tackle "iridium loss & regain (slightly more difficult)"

once above steps are complete we will close this PR here and open a new one

File: src/modules/commander/state_machine_helper.cpp [hunk 6]
```diff
@@ -834,48 +834,63 @@ void set_link_loss_nav_state(vehicle_status_s *status, actuator_armed_s *armed,
...
-		status->nav_state = vehicle_status_s::NAVIGATION_STATE_AUTO_RTL;
-
-	} else if (link_loss_act == link_loss_actions_t::AUTO_LAND && status_flags.condition_local_position_valid) {
-		status->nav_state = vehicle_status_s::NAVIGATION_STATE_AUTO_LAND;
-
-	} else if (link_loss_act == link_loss_actions_t::TERMINATE) {
-		status->nav_state = vehicle_status_s::NAVIGATION_STATE_TERMINATION;
-		armed->force_failsafe = true;
+	switch (link_loss_act) {
+	case (link_loss_actions_t::DISABLED):
```
Comment (MaEtUgR) [inline src/modules/commander/state_machine_helper.cpp]: Is this fallthough intended?
`DISABLED` does the same as `AUTO_RECOVER`?

File: src/modules/commander/state_machine_helper.cpp [hunk 7]
```diff
@@ -834,48 +834,63 @@ void set_link_loss_nav_state(vehicle_status_s *status, actuator_armed_s *armed,
...
-			status->nav_state = vehicle_status_s::NAVIGATION_STATE_AUTO_LANDGPSFAIL;
+			} else {
+				// TODO: FW position controller doesn't run without condition_global_position_valid
+				status->nav_state = vehicle_status_s::NAVIGATION_STATE_AUTO_LANDGPSFAIL;
+				return;
+			}
 		}
 
-	} else {
+	// FALLTHROUGH
```
Comment (MaEtUgR) [inline src/modules/commander/state_machine_helper.cpp]: only falls through for `rotary_wing` when neither position nor altitude are valid and never for fixed wing.

File: src/modules/commander/Commander.hpp (hunk 3 above)
Comment (MaEtUgR) [reply src/modules/commander/Commander.hpp]: Shouldn't all of these `uint64_t` be hrt_abstime? They are probably all used to compare to another such timestamp.

File: src/modules/commander/Commander.hpp [hunk 8]
```diff
@@ -155,27 +166,21 @@ class Commander : public ModuleBase<Commander>, public ModuleParams
 
 	void mission_init();
 
-	/**
-	 * Update the telemetry status and the corresponding status variables.
-	 * Perform system checks when new telemetry link connected.
-	 */
-	void poll_telemetry_status();
```
Comment (MaEtUgR) [inline src/modules/commander/Commander.hpp]: Is not contained in `data_link_check(...)` ✔️

File: msg/telemetry_status.msg [hunk 9]
```diff
@@ -1,18 +1,41 @@
-uint64 timestamp			# time since system start (microseconds)
+uint8 LINK_TYPE_GENERIC = 0
```
Comment (MaEtUgR) [inline msg/telemetry_status.msg]: Do these enums need to be in sync with any MAVLink definition? If yes could it be mentioned in a comment?

File: src/modules/commander/Commander.cpp (hunk 2 above)
Comment (cmic0) [reply src/modules/commander/Commander.cpp]: Done in [41b4e20](https://github.com/PX4/Firmware/pull/11550/commits/41b4e20d0f6bd66a4a1f464e82ce40242ddacd11).

**Comment by bkueng** [comment]
Replaced by #11550.
